consensus/src/pending_order_votes.rs

```diff
@@ -0,0 +1,262 @@
+    ErrorAggregatingSignature(VerifyError),
+}
+
+#[derive(Debug, PartialEq, Eq)]
+enum OrderVoteStatus {
+    EnoughVotes,
+    NotEnoughVotes,
+}
+
+/// A PendingVotes structure keep track of order votes for the last few rounds
+pub struct PendingOrderVotes {
+    /// Maps LedgerInfo digest to associated signatures (contained in a partial LedgerInfoWithSignatures).
+    /// Order vote status stores caches the information on whether the votes are enough to form a QC.
+    li_digest_to_votes: HashMap<
+        HashValue, /* LedgerInfo digest */
+        (OrderVoteStatus, LedgerInfoWithPartialSignatures),
+    >,
+    /// Map of (Author, round number) to order vote. This is useful to discard multiple votes.
+    author_to_order_vote: HashMap<(Author, u64), OrderVote>,
+}
+
+impl PendingOrderVotes {
+    /// Creates an empty PendingOrderVotes structure
+    pub fn new() -> Self {
+        Self {
+            li_digest_to_votes: HashMap::new(),
+            author_to_order_vote: HashMap::new(),
+        }
+    }
+
+    /// Add a vote to the pending votes
+    // TODO: Should we add any counters here?
+    pub fn insert_order_vote(
+        &mut self,
+        order_vote: &OrderVote,
+        validator_verifier: &ValidatorVerifier,
+    ) -> OrderVoteReceptionResult {
+        // derive data from vote
+        let li_digest = order_vote.ledger_info().hash();
+        let round = order_vote.ledger_info().round();
+
+        if let Some(previously_seen_vote) =
+            self.author_to_order_vote.get(&(order_vote.author(), round))
+        {
+            // is it the same vote?
+            if li_digest == previously_seen_vote.ledger_info().hash() {
+                return OrderVoteReceptionResult::DuplicateVote;
+            } else {
+                // we have seen a different vote for the same round
+                error!(
+                    SecurityEvent::ConsensusEquivocatingOrderVote,
+                    remote_peer = order_vote.author(),
+                    order_vote = order_vote,
+                    previous_vote = previously_seen_vote
+                );
+
+                return OrderVoteReceptionResult::EquivocateVote;
+            }
+        }
+
+        self.author_to_order_vote
+            .insert((order_vote.author(), round), order_vote.clone());
+
+        // obtain the ledger info with signatures associated to the order vote's ledger info
+        let (status, li_with_sig) = self.li_digest_to_votes.entry(li_digest).or_insert_with(|| {
+            // if the ledger info with signatures doesn't exist yet, create it
+            (
+                OrderVoteStatus::NotEnoughVotes,
+                LedgerInfoWithPartialSignatures::new(
+                    order_vote.ledger_info().clone(),
+                    PartialSignatures::empty(),
+                ),
+            )
+        });
+        let validator_voting_power = validator_verifier
+            .get_voting_power(&order_vote.author())
+            .unwrap_or(0);
```

Comment:
we should not unwrap but just return error with unknown author

Reply:
Done.